### DISCUSSION TITLE
Laser pointer

### DISCUSSION THREAD

**Discussion by K4LCIFER** [xournalpp/xournalpp#2643] [Ideas]
The app "goodnotes" has a really underrated laser pointer pen feature. Wen you write with it, it will make a temporary red stroke that fades after a second or two. This comes in really handy when you're trying to annotate/explain something to someone in a live scenario. You can draw temporary annotations to draw attention. I think it would be great if this were added to Xournal++ as a another pen option.

**Comment by Technius** [comment]
You may be interested in #1778.

**Reply by K4LCIFER** [reply]
Do you have an update on the status? Some of the conversations that I have seen about this were opened quite a number of months ago, and don't really seem to have been touched since. I am eagerly awaiting this feature!

**Reply by LittleHuba** [reply]
We are unfortunately a small team that does this work in their free time. Therefore it is unfortunately very common that such tickets take some months to gather speed. If you have experience in programming you are welcome to provide a PR with this feature.

**Reply by StefanL38** [reply]
High Hubert,
I have quite some programming experience with delphi and C++ for Arduino. So I'm familiar with the basic concepts. What does "provide a PR" mean? Pull-Request? I'm not even much familiar with GitHUB. So from knowing almost nothing about Pull-Requests my imagination of it is: at first take 50 hours of time to learn how to install and use GIT on my local computer.  But maybe I'm totally wrong. 

So if you can show me the place where in the source-code I would have to insert a laser-pointer-feature
I would take time to write it. Because I would love to have the laserpointer-feature for live-sesions too.  

If the main source code is not well documented my opinion is: 
it will be easier for me to write an autohotkey-script that will 
- activate lasermode 
- switching to highlighter-pen 
- send a Ctrl-z after a useradjustable delay 

than digging into a non-documented sourcecode. 
best regards Stefan

**Reply by rolandlo** [reply]
Yes, PR does mean Pull-Request. Getting started with Git is not that hard. Following [this tutorial](https://firstcontributions.github.io/) it is just 5 minutes to make your first PR. 
About the place in the source-code: Take a look at PR #1822 and the last 4 commits in [this implementation](https://github.com/michaelcadilhac/xournalpp/commits/laser-tool)

**Reply by StefanL38** [reply]
Hi rolando,

thank you for answering.

Just what I estimated. I have to install Git on my local computer I have to clone some repository create a branch
doing all this using a command-line tool. 
This tutorial is 5 minutes fast because it does NOT explain what it is doing in detail. 
And it would be much easier to understand if an example would be used to explain what is going on.
With example I mean something like this:

Imagine somebody has written a manual how to draw santa-clauses house (picture of what is meant)
This house just a rectangle with a triangle as roof and two diagonals through the rectangle)
The explanation above if so elementary that you REALLY can expect that it can be understood by any person 
that has reached age of 14 and was at school)

The author of this manuals has used letters for each edge-point.
The manual is stored into file    ......
You want to change this manual to use numbers for each edge-point instead of letters.

Each GitHub-project has some standard additional files for organasation. These files are
....  their function is .....
Then some pictures that explain what structure the project has on github.com
what of this structure is copied to my local harddisk
what is again added as folders and subfolders if I create a fork

then open file .... edit the text and the picture to use numbers instead of letters save file 
and then do ....

Of course with explaining all this it is not done in five minutes.
But I'm GUIDED for maybe 30 minutes instead of beeing lost alone for hours and days 
 
That's why I estimate that it needs 50 hours to get enough familiar with it to be able to re-solve an error myself.
Without that knowledge each error escalates into its own 10 ask&answer-thread. 

I call this "**the expert's blindness for beginners difficulties**"  You - the expert know so much about it that you no longer recognise how much knowledge is really common for you.

You show this blindness by writing 
About the place in the source-code: Take a look at PR #1822 and the last 4 commits in this implementation

This arises the following newbee-questions because I **AM** a newbee about GitHUB
what is a commit?  And what do you mean by "implementation" in this case?
And most important **WHERE** is the source-code?? What I can see at this link is a pretty long discussion but no source-code.

best regards Stefan

**Comment by rolandlo** [comment]
Sorry for showing expert's blindness. I don't mind answering follow-up questions though.

First of all, if you `clone` a code repository (by executing `git clone` followed by the URL of the forked repository) a new folder named after the repository is created on your computer and all files of the repository (consisting of source files, build scripts, various resource files etc.) are copied into that folder. You can now build the project on your computer, modify the files in the repository and code up a laser feature for example. Once you are finished and want to get feedback you can `git push` your changes to your forked remote repository and create a pull request to the original repository. 

Take a look here:
![grafik](https://user-images.githubusercontent.com/40485433/108267615-bb22ee00-716b-11eb-9ddc-2ad8248c6a27.png)

This is a Pull Request by one of the maintainers @Technius, which aims (or aimed) to add a simple laser tool to Xournal++. There are tabs for `Conversation`, `Commits`, `Checks` and `Files changed`

Take a look at the `Files changed` tab. Clicking on it (after following the link #1822) you will see which files were modified (6 in total as you can see from the image). That should answer your question about "where in the source-code I would have to insert a laser-pointer-feature"

In the `Conversation` tab people discuss the changes to the repository proposed in the pull request. In this case the general consensus was that it was not quite what we were looking for. 

If you create a pull request, you first make one `commit`. This is just a set of changes made to the repository that you contribute in one go. Later you may want to improve your pull request and add more commits. 

There is also a `Checks` tab. After every commit the project is automatically been built online (on Azure Pipelines), so we can see if it builds correctly on various operating systems (Windows, MacOS and Linux) and if the code is formatted in the right way. 

I hope this helps you a step further. Feel free to ask more questions.

**Reply by StefanL38** [reply]
Hi Roland,

thank you for answering.

I read through the code you linked to with #1822. After reading the code there I need your opinion after describing my 
knowledge about programming and describing my idea how to contribute through writing code.

I have quite a lot experience with delphi. I have written GUI-based control-software for industrial machines creating / adapating the GUI, connecting to various kinds of data-exchange mechanisms, and all the control-algorythms including complex state-machines,  I have a rather small knowledge about C++ for Arduino.  

To give you an example:
a wireless dataexchange-protocol called esp-now 
you include the esp-now-library
`#include <ESP8266WiFi.h> // for use with ESP32 use file "WiFi.h"
#include <espnow.h>      // for use with ESP32  use file "esp_now.h" (underline in the middle)
`

register a callback-function to the esp-now library

`  esp_now_register_recv_cb(OnDataRecv);

// Callback function that will be executed when data is received
void OnDataRecv(uint8_t * mac, uint8_t *incomingData, uint8_t NoOfBytesRcv) 
{ 
  // copy data bytewise from variable incomingData to variable myData
  memcpy(&my_received_ESP_NOW_Data, incomingData, sizeof(my_received_ESP_NOW_Data));
  //these lines must match the variables inside the ESP_NOW-data-structure
  Serial.print("No of Bytes received: ");
  Serial.println(NoOfBytesRcv);
  Serial.print("Array of Char: ");
  Serial.println(my_received_ESP_NOW_Data.MyESP_NOW_MsgStr);
  Serial.print("Int: ");
  Serial.println(my_received_ESP_NOW_Data.MyESP_NOW_Int);

  Serial.println();
}
`
and communication can start = the coding for receiving data is done
pretty easy to do and there are examples on the internet that show the important steps

what is way beyond my knowledge about C++ are things like that:
`PresentationLaser(GtkWidget* widget);
    PresentationLaser(const PresentationLaser&) = delete;
    PresentationLaser(const PresentationLaser&&) = delete;
    virtual ~PresentationLaser();
`
what does the "&"  and the "&&" behind   PresentationLaser mean?
what does the "~" in front of function PresentationLaser() mean?
what does the controlword "virtual" change?

and
`if (auto* laser = this->inputContext->getView()->getPresentationLaser()) {`
what does the "this->"    do???  and in this case it is even used multiple times  sequentially

You might think    "OMG  looong way to learn C/C++!"     I  guess this is true.    If it is true clearly state it.
my imagination of how I could contribute to xournal++ is:

somebody shows me the exact place where:
- to add the defintion for the button "LaserMarker"

- I have to add the eventhandler "ClickOnLaserMarker"  

and as a line drawn with the LaserMarker should vanish after 1-5 seconds where to put the code inside "main-loop" 
or how to setup a timer that gets executed once every 100 milliseconds for increasing the transparency.
and "refreshing" the picture on the screen.
that "undraws" the line through increasing transparency of the color and if reached 100% transparency delete the whole line that has been drawn like a Ctrl-z would undo a linedraw action.

So the main thing is showing all this by pointing me to the places where in the code the button Highlighter is defined 
definition of Eventhandler ClickOnHighLightButton etc. is.

I have three Thinkpads laying around for experiments. 
If you can create a portable version of the whole toolchain (with adjusted makefile and whatever else is needed) that is able to re-compile the xournal++ source-code  and pack this portable toolchain into a zipfile so that "installing" the tool-chain is done just by unzipping the zip-file I could start writing code.

This is an approach of start real coding through preparations 
- and learn the details about the coding by small changes to existing code (the existing highlighter-functionality) 
- learn how the details of the tool-chain work from using it starting from a ready to run and working example instead of 
hassling around with all the missing details of poor and too short documentation.

I have no idea how much time this preparation needs.
and I have no idea if it is even possible to setup a working toolchain  in a 100% portable way. That just unzipping will be sufficient to make it work.

If all this could be done within one hour by you it is really worth a try. If the initial steps need much more time  for you 
I'm out contributing on this path.

best regards Stefan

**Reply by rolandlo** [reply]
Hi Stefan

I must apologize, but it seems to me that your C++ knowledge is currently not sufficient to do the job. I would spend more time explaining you the details than doing the job myself. 

Let me still explain you one of your questions:

In C++ we use a lot of pointers. A pointer stores the memory address of the object it points to. If a pointer points to a structure, union or class, you can use the arrow operator `->` to access its members and member functions. In the example you mentioned `this` is a pointer to the current object, an instance of the class `PenInputHandler`. The `PenInputHandler` class is derived from the `AbstractInputHandler` class and the latter has a pointer `InputContext* inputContext;` To access this object we use `this->inputContext`. Now the `InputContext` class has a pointer `XournalView* getView();`. To access this object we use `this->inputContext->getView()`. That's how we can combine multiple arrow operators.

 For a full explanation of the arrow operator see [this link](https://www.geeksforgeeks.org/arrow-operator-in-c-c-with-examples/)

Getting started with a large open source project like Xournal++ means reading a lot of code, reading about things you don't understand in external sources, and starting to manipulate the code by solving simple issues. To add a laser tool in the GUI you would have to change about 10 files, I guess. The best way to find out about it would be to read the code of a pull request, which also introduced a new button.

Still thank you for your will to contribute. Maybe you will come back one day with a stronger C++ background.

Best regards, Roland
